instrumentation/opentelemetry-instrumentation-urllib/tests/test_urllib_integration.py

```diff
@@ -0,0 +1,382 @@
+            span.attributes,
+            {
+                "component": "http",
+                "http.method": "GET",
+                "http.url": self.URL,
+                "http.status_code": 200,
+                "http.status_text": "OK",
+                "http.response.body": "Hello!",
+            },
+        )
+
+    def test_custom_tracer_provider(self):
+        resource = resources.Resource.create({})
+        result = self.create_tracer_provider(resource=resource)
+        tracer_provider, exporter = result
+        URLLibInstrumentor().uninstrument()
+        URLLibInstrumentor().instrument(tracer_provider=tracer_provider)
+
+        result = self.perform_request(self.URL)
+        self.assertEqual(result.read(), b"Hello!")
+
+        span = self.assert_span(exporter=exporter)
+        self.assertIs(span.resource, resource)
+
+    def test_requests_exception_with_response(self, *_, **__):
+
+        with self.assertRaises(HTTPError):
+            self.perform_request("http://httpbin.org/status/500")
+
+        span = self.assert_span()
+        self.assertEqual(
+            dict(span.attributes),
+            {
+                "component": "http",
+                "http.method": "GET",
+                "http.url": "http://httpbin.org/status/500",
+                "http.status_code": 500,
+                "http.status_text": "Internal Server Error",
+            },
+        )
+        self.assertEqual(span.status.status_code, StatusCode.ERROR)
+        self.assertIsNotNone(URLLibInstrumentor().meter)
+        self.assertEqual(len(URLLibInstrumentor().meter.metrics), 1)
+        recorder = URLLibInstrumentor().meter.metrics.pop()
+        match_key = get_dict_as_key(
+            {
+                "http.method": "GET",
+                "http.status_code": "500",
+                "http.url": "http://httpbin.org/status/500",
+                "http.flavor": "1.1",
+            }
+        )
+        for key in recorder.bound_instruments.keys():
+            self.assertEqual(key, match_key)
+            # pylint: disable=protected-access
+            bound = recorder.bound_instruments.get(key)
+            for view_data in bound.view_datas:
+                self.assertEqual(view_data.labels, key)
+                self.assertEqual(view_data.aggregator.current.count, 1)
+
+    def test_requests_basic_exception(self, *_, **__):
+        with self.assertRaises(Exception):
+            self.perform_request(self.URL_EXCEPTION)
+
+        span = self.assert_span()
+        self.assertEqual(span.status.status_code, StatusCode.ERROR)
+
+    def test_requests_timeout_exception(self, *_, **__):
+        with self.assertRaises(Exception):
+            opener = urllib.request.build_opener()
+            opener.open(self.URL_TIMEOUT, timeout=0.0001)
+
+        span = self.assert_span()
+        self.assertEqual(span.status.status_code, StatusCode.ERROR)
+
+
+class TestRequestsIntegration(RequestsIntegrationTestBase, TestBase):
```

Comment:
just out of curiosity: do you expect that we will exercise the integration in more than just the urllib.request? Curious why you followed this pattern of factoring out shared validation and allowing one to override how the initial request is formed.

Reply:
I'm not sure that I'm going to extend instrumentation of `urllib` by adding something out from `urllib.request`.
But who knows maybe someone will want to add something that I've missed.
Basically I'd just followed the test patterns that were already written in repo.

Reply:
okay, sounds good. It's a bit confusing to add this type of structure without a clear use for it, since we can always refactor.

But not a big deal either way.